main.js

```diff
@@ -1,60 +1,59 @@
-  {text: 'Human sacrifice! Dogs and cats living together. Mass hysteria!', movie: 'Ghostbusters'}
-];
-const displayArea = document.getElementById('display-area');
-
-bindEvents();
-displayNewQuote();
-
-function bindEvents() {
-  const newQuoteButton = document.getElementById('new-quote-button');
-  newQuoteButton.addEventListener('click', handleNewQuoteClick);
-}
-
-function displayNewQuote() {
-  const quotePara = document.getElementById('quote');
-  const refCite = document.getElementById('ref');
-  const quote = getRandomQuote(quotes);
-
-  quotePara.innerText = '\“' + quote['text'] + '\”';
-  refCite.innerText = quote['movie'];
-  displayArea.classList.toggle('show');
-  updateTwitterIntent(quote);
-};
-
-function getRandomIndex(array) {
-  return Math.floor(Math.random() * array.length);
-}
-
-function getRandomQuote(quotes) {
-  const index = getRandomIndex(quotes);
-  return quotes[index];
-}
-
-function handleNewQuoteClick() {
-  displayArea.classList.toggle('show');
-  setTimeout(displayNewQuote, 350);
-}
-
-function updateTwitterIntent(quote) {
-  const twitterLink = document.getElementById('tweet-quote-link');
-  const intentAddress = `https://twitter.com/intent/tweet?text=“${quote.text}”  — ${quote.movie}`;
-  twitterLink.setAttribute('href', intentAddress);
+fetch('quotes.json')
+  .then(response => {
+    if (!response.ok) {
+      document.getElementById('quote-box').textContent = `
+        Network request failed with status 
+        ${response.status}: ${response.statusText}. 
+        Please refresh to try again.
+      `;
+    }
+    return response.json();
+  })
+  .then(json => {
+    const quotes = json;
+    init(quotes);
+  });
+
+function init(quotes) {
+  const displayArea = document.getElementById('display-area');
+  bindEvents();
+  displayNewQuote();
+
+  function bindEvents() {
+    const newQuoteButton = document.getElementById('new-quote-button');
+    newQuoteButton.addEventListener('click', handleNewQuoteClick);
+  }
+
+  function displayNewQuote() {
+    const quotePara = document.getElementById('quote');
+    const refCite = document.getElementById('ref');
+    const refYear = document.getElementById('ref-year');
+    const quote = getRandomQuote(quotes);
+
+    quotePara.innerText = `“${quote.text}”`;
+    refCite.innerText = quote.movie;
+    refYear.innerText = ` (${quote.year})`;
+    displayArea.classList.toggle('show');
+    updateTwitterIntent(quote);
+  };
+
+  function getRandomIndex(array) {
+    return Math.floor(Math.random() * array.length);
+  }
+
+  function getRandomQuote(quotes) {
+    const index = getRandomIndex(quotes);
+    return quotes[index];
+  }
+
+  function handleNewQuoteClick() {
+    displayArea.classList.toggle('show');
+    setTimeout(displayNewQuote, 350);
+  }
+
+  function updateTwitterIntent(quote) {
+    const twitterLink = document.getElementById('tweet-quote-link');
+    const intentAddress = `https://twitter.com/intent/tweet?text=“${quote.text}”  — ${quote.movie}`;
```

Comment:
I am pretty sure `“` is not allowed in a URL. most browsers will be forgiving though. research: so what characters are valid in URLs? how do you escape letters that are not allowed?

Reply:
Thanks for the heads-up. 

If I understand correctly, unreserved unicode characters like `“` `”` `‘` `’` do not need to be encoded in URLs in HTML because "_browsers will take care of them per the URL parsing algorithm in the HTML spec_" (I got this from the **Update** comment from Mathias Bynens highlighted in yellow [here](https://perishablepress.com/stop-using-unsafe-characters-in-urls/#html-javascript)). So it sounds to me like `https://twitter.com/intent/tweet?text=“This is a quote”` is a valid and safe URL when it is assigned as the value of the `href` attribute on an anchor element in HTML.

This appears to be true for Chrome and Firefox, although both seem to "hide" the encoding in the address bar. I copy+pasted the URLs from the browser into a plain text editor, and this is more or less what you see: `https://twitter.com/intent/tweet?text=%E2%80%9CThis is a quote%E2%80%9D%`.

However, the quotes themselves do contain a number of reserved characters: `!` `?` `,` (e.g. "You talkin’ to me?", "You can't handle the truth!"). Even though the twitter intent links work, I imagine it would be a good idea to percent encode these characters anyways.

Reply:
yep. I agree. if in doubt make sure to encode them.